src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/NextReviewDateUpdaterService.kt

```diff
@@ -0,0 +1,86 @@
+import uk.gov.justice.digital.hmpps.incentivesapi.dto.prisonapi.IepLevel
+import uk.gov.justice.digital.hmpps.incentivesapi.jpa.NextReviewDate
+import uk.gov.justice.digital.hmpps.incentivesapi.jpa.PrisonerIepLevel
+import uk.gov.justice.digital.hmpps.incentivesapi.jpa.repository.NextReviewDateRepository
+import uk.gov.justice.digital.hmpps.incentivesapi.jpa.repository.PrisonerIepLevelRepository
+import java.time.LocalDate
+
+/**
+ * Recalculates and persists next review dates
+ */
+@Service
+class NextReviewDateUpdaterService(
+  private val prisonerIepLevelRepository: PrisonerIepLevelRepository,
+  private val nextReviewDateRepository: NextReviewDateRepository,
+  private val prisonApiService: PrisonApiService,
+  private val offenderSearchService: OffenderSearchService,
+) {
+
+  /**
+   * Update next review date for the given bookingId
+   *
+   * @param bookingId of the offender to update
+   *
+   * @return the nextReviewDate for the given bookingId
+   * */
+  suspend fun update(bookingId: Long): LocalDate {
+    val locationInfo = prisonApiService.getPrisonerInfo(bookingId, useClientCredentials = true)
+    val prisonerNumber = locationInfo.offenderNo
+    val offender = offenderSearchService.getOffender(prisonerNumber)
+
+    return updateMany(listOf(offender))[offender.bookingId]!!
+  }
+
+  /**
+   * Update next review date for the given offenders
+   *
+   * @param offenders is the list of offenders to update
+   *
+   * @return a map with bookingIds as keys and nextReviewDate as value
+   * */
+  suspend fun updateMany(offenders: List<OffenderSearchPrisoner>): Map<Long, LocalDate> {
+    if (offenders.isEmpty()) {
+      return emptyMap()
+    }
+
+    val offendersMap = offenders.associateBy(OffenderSearchPrisoner::bookingId)
+    val bookingIds = offendersMap.keys.toList()
+
+    val iepLevels: Map<String, IepLevel> = prisonApiService.getIncentiveLevels()
+
+    // NOTE: This is to account for bookingIds potentially without any review record
+    val bookingIdsNoReviews = bookingIds.associateWith { emptyList<PrisonerIepLevel>() }
+    val reviewsMap = bookingIdsNoReviews + prisonerIepLevelRepository.findAllByBookingIdInOrderByReviewTimeDesc(bookingIds)
+      .toList()
+      .groupBy(PrisonerIepLevel::bookingId)
+
+    val nextReviewDateRecords = reviewsMap.map {
+      val bookingId = it.key
+      val iepDetails = it.value.toIepDetails(iepLevels)
+      val offender = offendersMap[bookingId]!!
+
+      val nextReviewDate = NextReviewDateService(
+        NextReviewDateInput(
+          dateOfBirth = offender.dateOfBirth,
+          receptionDate = offender.receptionDate,
+          hasAcctOpen = offender.acctOpen,
+          iepDetails = iepDetails,
+        )
+      ).calculate()
+
+      NextReviewDate(
+        bookingId = bookingId,
+        nextReviewDate = nextReviewDate,
+      )
+    }
+
+    nextReviewDateRepository.saveAll(nextReviewDateRecords)
```

Comment:
tho the docs are extremely sparse, maybe `nextReviewDateRecords = nextReviewDateRepository.saveAll(nextReviewDateRecords).toList()` could pick up things like updated `when_created`. side benefit is that `toList` collects.

Reply:
👌 however in this context I don't really care about the result of `saveAll()` (aside from the fact it actually _saves_ things! 😝) so `collect()` should do :)

Reply:
Alright, I'm changing this and the tests to use the returned flow 👍

Reply:
Done here: 7622c201d99891938233b5af56ab8507752b2397